sdk/identity/identity/src/msal/msalClient.ts

```diff
@@ -0,0 +1,260 @@
+ */
+export function generateMsalConfiguration(
+  clientId: string,
+  tenantId: string,
+  createMsalClientOptions: CreateMsalClientOptions = {},
+): msal.Configuration {
+  const resolvedTenant = resolveTenantId(msalLogger, tenantId, clientId);
+
+  // TODO: move and reuse getIdentityClientAuthorityHost
+  const authority = getAuthority(
+    resolvedTenant,
+    createMsalClientOptions.authorityHost ?? process.env.AZURE_AUTHORITY_HOST,
+  );
+
+  const httpClient = new IdentityClient({
+    ...createMsalClientOptions.tokenCredentialOptions,
+    authorityHost: authority,
+    loggingOptions: createMsalClientOptions.loggingOptions,
+  });
+
+  const msalConfig: msal.Configuration = {
+    auth: {
+      clientId,
+      authority,
+      knownAuthorities: getKnownAuthorities(
+        resolvedTenant,
+        authority,
+        createMsalClientOptions.disableInstanceDiscovery,
+      ),
+    },
+    system: {
+      networkClient: httpClient,
+      loggerOptions: {
+        loggerCallback: defaultLoggerCallback(createMsalClientOptions.logger ?? msalLogger),
+        logLevel: getMSALLogLevel(getLogLevel()),
+        piiLoggingEnabled: createMsalClientOptions.loggingOptions?.enableUnsafeSupportLogging,
+      },
+    },
+  };
+  return msalConfig;
+}
+
+/**
+ * Represents the state necessary for the MSAL (Microsoft Authentication Library) client to operate.
+ * This includes the MSAL configuration, cached account information, Azure region, and a flag to disable automatic authentication.
+ *
+ * @internal
+ */
+interface MsalClientState {
+  /** The configuration for the MSAL client. */
+  msalConfig: msal.Configuration;
+
+  /** The cached account information, or null if no account information is cached. */
+  cachedAccount: msal.AccountInfo | null;
+}
+
+/**
+ * Creates an instance of the MSAL (Microsoft Authentication Library) client.
+ *
+ * @param clientId - The client ID of the application.
+ * @param tenantId - The tenant ID of the Azure Active Directory.
+ * @param createMsalClientOptions - Optional. Additional options for creating the MSAL client.
+ * @returns An instance of the MSAL client.
+ *
+ * @public
+ */
+export function createMsalClient(
+  clientId: string,
+  tenantId: string,
+  createMsalClientOptions: CreateMsalClientOptions = {},
+): MsalClient {
+  const state: MsalClientState = {
+    msalConfig: generateMsalConfiguration(clientId, tenantId, createMsalClientOptions),
+    cachedAccount: createMsalClientOptions.authenticationRecord
+      ? publicToMsal(createMsalClientOptions.authenticationRecord)
+      : null,
+  };
+
+  // configure persistence
+  // configure broker
+
+  let confidentialApp: msal.ConfidentialClientApplication | undefined = undefined;
+  async function getConfidentialApp(
+    _options: GetTokenOptions = {},
+  ): Promise<msal.ConfidentialClientApplication> {
+    // abort requests
+
+    if (confidentialApp === undefined) {
+      // TODOs:
+      // CAE / non-CAE
```

Comment:
Each of these will be broken out to separate issues if we decide to move forward